crates/sui-graphql-rpc/src/types/query.rs

```diff
@@ -62,9 +67,141 @@ impl Query {
     }
 
     // availableRange - pending impl. on IndexerV2
-    // dryRunTransactionBlock
     // coinMetadata
 
+    /// Simulate running a transaction to inspect its effects without
+    /// committing to them on-chain.
+    ///
+    /// `txBytes` either a `TransactionData` struct or a `TransactionKind`
+    ///     struct, BCS-encoded and then Base64-encoded.  The expected
+    ///     type is controlled by the presence or absence of `txMeta`: If
+    ///     present, `txBytes` is assumed to be a `TransactionKind`, if
+    ///     absent, then `TransactionData`.
+    ///
+    /// `txMeta` the data that is missing from a `TransactionKind` to make
+    ///     a `TransactionData` (sender address and gas information).  All
+    ///     its fields are nullable: `sender` defaults to `0x0`, if
+    ///     `gasObjects` is not present, or is an empty list, it is
+    ///     substituted with a mock Coin object, `gasPrice` defaults to
+    ///     the reference gas price, `gasBudget` defaults to the max gas budget
+    ///     and `gasSponsor` defaults to the sender.
+    ///
+    /// `skipChecks` optional flag to disable the usual verification
+    ///     checks that prevent access to objects that are owned by
+    ///     addresses other than the sender, and calling non-public,
+    ///     non-entry functions, and some other checks.  Defaults to false.
+    async fn dry_run_transaction_block(
+        &self,
+        ctx: &Context<'_>,
+        tx_bytes: String,
+        tx_meta: Option<TransactionMetadata>,
+        skip_checks: Option<bool>,
+    ) -> Result<DryRunResult> {
+        let skip_checks = skip_checks.unwrap_or(false);
+
+        let sui_sdk_client: &Option<SuiClient> = ctx
+            .data()
+            .map_err(|_| Error::Internal("Unable to fetch Sui SDK client".to_string()))
+            .extend()?;
+        let sui_sdk_client = sui_sdk_client
+            .as_ref()
+            .ok_or_else(|| Error::Internal("Sui SDK client not initialized".to_string()))
+            .extend()?;
```

Comment:
It's interesting to me that we need both these checks -- I would have thought we either got a hold of the SDK and it was initialized, or we couldn't and we got `None` from the first step -- what's the situation in which the first step would fail if not for the fact that `SuiClient` was not initialized?

Reply:
I'm also not too sure. This was in @oxade's original PR, which may not have been thought through. I guess we can just grab a `SuiClient` directly instead of  `Option<SuiClient>`.

Reply:
Hmm I tried and did get an error this way so I guess this double checking is needed? But I don't know why

Reply:
It sounds like we are inserting an `Option<SuiClient>` instead of just a `SuiClient` somewhere, and that's what could be cleaned up?

Reply:
Looks like this is the culprit:

https://github.com/MystenLabs/sui/blob/86857b2077314e424e471694f97ed76d93560ff2/crates/sui-graphql-rpc/src/server/builder.rs#L205-L210

This does not need to be an optional -- either we got the fullnode URL and we add the non-optional client as context data, or we didn't and we don't (or at this point, we could panic if we wanted to).

Reply:
I see. I think this can still be an option to allow for gql servers that don't want to serve dry run/execution requests.

Reply:
Indeed, it's reasonable to allow the the service to run without an SDK client set-up, but that just means we shouldn't panic if we are not given a fullnode URL. I still don't think we should add an `Option<SuiClient>` to the context, we can just add a `SuiClient` and rely on the fact that we can do a checked fetch for `SuiClient` from the context to see whether it is set or not.  In fact, we are already doing that today, that's what the first check is.